web/src/stores/clips.ts

```diff
@@ -53,93 +115,83 @@ export namespace Clips {
 
   export type Action =
     | LoadAction
+    | LoadErrorAction
     | RefillCacheAction
     | RemoveClipAction
     | AchievementAction;
 
   export const actions = {
-    refillCache: () => async (
-      dispatch: Dispatch<RefillCacheAction | LoadAction>,
-      getState: () => StateTree
-    ) => {
-      const state = getState();
-      if (localeClips(state).clips.length > MIN_CACHE_SIZE) {
-        return;
-      }
+    refillCache:
+      () =>
+      async (
+        dispatch: Dispatch<RefillCacheAction | LoadAction | LoadErrorAction>,
+        getState: () => StateTree
+      ) => {
+        const state = getState();
 
-      try {
-        dispatch({ type: ActionType.LOAD });
-        const clips = await state.api.fetchRandomClips(
-          MIN_CACHE_SIZE - localeClips(state).clips.length
-        );
+        // don't load if no contributable locale
+        if (!contributableLocales.includes(state.locale)) {
+          return;
+        }
+
+        if (localeClips(state).clips.length > MIN_CACHE_SIZE) {
+          return;
+        }
+
+        try {
+          dispatch({ type: ActionType.LOAD });
+
+          const randomClips = await state.api.fetchRandomClips(
+            MIN_CACHE_SIZE - localeClips(state).clips.length
+          );
+
+          const clips = await removeClipsWithErrors(randomClips);
 
-        dispatch({
-          type: ActionType.REFILL_CACHE,
-          clips: clips.map(clip => {
-            let sentence = clip.sentence;
-            try {
-              sentence.text = decodeURIComponent(sentence.text);
-            } catch (e) {
-              if (e.name !== 'URIError') {
-                throw e;
-              }
-            }
-
-            return {
-              id: clip.id,
-              glob: clip.glob,
-              sentence,
-              audioSrc: clip.audioSrc,
-            };
-          }),
-        });
-        await Promise.all(clips.map(({ audioSrc }) => fetch(audioSrc)));
-      } catch (err) {
-        if (err instanceof XMLHttpRequest) {
```

Comment:
I think the idea with this is if one of clips was a bit dodge we would just request a whole new batch of clips but I think this could end up with us in a infinite loop if AWS is down?

Reply:
No, `bucket.ts` has provisions that checks clips one by one, so that if one of the clips is a dodge it'll try to pick up a new one, up to a 1.5x multiplier of the default number of clips required: https://github.com/common-voice/common-voice/blob/main/server/src/lib/bucket.ts#L122-L124 (I'm not sure if that explanation makes sense, the code is probably clearer)

If AWS is down across the board and none of the clips are returning, that would be a different problem - there should probably be some max # of retries?

Reply:
Do you know why we'd expect XMLHttpRequest errors? My hunch is if it errors we should tell the user and they can come back/refresh

Reply:
I don't, sorry.